conbench/entities/distribution.py

```diff
@@ -108,109 +111,160 @@ def get_distribution(benchmark_result, limit):
 
-    parent = None
-    if result:
-        commit_id = result[1]
-        parent = Commit.get(commit_id)
+    # see Commit.ancestor_commit_query() comments for why we order this way
+    closest_benchmark_result = query.order_by(
+        "branch_order", ancestor_commits.c.ancestor_timestamp.desc()
+    ).first()
 
-    return parent
+    if not closest_benchmark_result:
+        log.debug(
+            "Couldn't find closest ancestor: there were no matching BenchmarkResults"
+        )
+        return None
 
+    return closest_benchmark_result.run.commit
 
-def set_z_scores(benchmark_results):
-    if not benchmark_results:
-        return
 
-    for benchmark_result in benchmark_results:
-        benchmark_result.z_score = None
+def set_z_scores(benchmark_results: List["BenchmarkResult"]):
+    """Populate the "z_score" attribute of each given BenchmarkResult, comparing it to
+    the distribution of BenchmarkResults that share the same hardware/case/context,
+    ending with the most recent BenchmarkResult in the given BenchmarkResult's commit
+    ancestry on the default branch.
 
-    first = benchmark_results[0]
-    parent_commit = get_closest_parent(first.run)
-    if not parent_commit:
-        return
+    We choose a commit from the default branch as the end commit, because we probably
+    don't want "bad" commits early on in a PR to set artificially low z-scores later in
+    the PR.
 
-    where = [
-        Distribution.commit_id == parent_commit.id,
-        Distribution.hardware_hash == first.run.hardware.hash,
-    ]
-    if len(benchmark_results) == 1:
-        where.extend(
-            [
-                Distribution.case_id == first.case_id,
-                Distribution.context_id == first.context_id,
-            ]
-        )
+    For the start commit of the distribution, uses whatever commit_limit was used when
+    writing to the Distribution table.
+
+    Should not raise anything. If we can't find a z-score for some reason, the z_score
+    attribute will be None on that BenchmarkResult.
+    """
+    if not benchmark_results:
+        return
 
-    cols = [
-        Distribution.case_id,
-        Distribution.context_id,
-        Distribution.mean_mean,
-        Distribution.mean_sd,
-    ]
+    # Find the default branch first. If we can't find it, we'll just find the closest
+    # ancestor commit to each BenchmarkResult regardless of branch, which isn't ideal
+    # (see docstring) but it's good enough.
+    a_default_commit = benchmark_results[0].run.commit.get_fork_point_commit()
+    default_branch = a_default_commit.branch if a_default_commit else None
 
-    distributions = Session.query(*cols).filter(*where).all()
-    lookup = {f"{d.case_id}-{d.context_id}": d for d in distributions}
+    # Cache Distribution query results by commit/hardware/case/context, in case multiple
+    # input benchmark_results (e.g. from the same Run) share those attributes
+    cached_distributions = {}
```

Comment:
The stupid saying is: there's only two hard problems in ...
- off-by one
- naming
- cache invalidation -- haha

So. When/how is this cache invalidated? I was `Ctrl + F` this patch to find the word "invalidation" but didn't find it :)

Reply:
In this case, the cache isn't referenced anymore after exiting this function scope (so it's eventually garbage-collected). So the cache won't be invalidated during a call to `set_z_scores()`. 

I don't think we need to worry about one of these query results changing during one call to `set_z_scores()`, because:

- that only would happen when BenchmarkResults are being POSTed at the exact same time they're being GETed (probably rare)
- even if it does happen, in most cases the z-score might only be changed by a few decimal places (since theoretically, the distribution mean/standard deviation don't have big changes too often)